### PR TITLE
[clang][tmo] Make TMO inference more robust

### PR DESCRIPTION
**Pull request by ojhunt**
A wide range of changes to make TMO inference more robust.

The biggest change is accepting that at codegen we just cannot guarantee that a call with the TMO attribute has had inference performed, due to a combination of PCH, modules, and template instantiation order. We could in principle perform last minute inference at that point but historically that has caused problems due to what is fundamentally a Sema operation happening after Sema. This may be something we should return to in future but for now it does not seem like a fantastic idea.

Alternatively we would need to perform a pass at the end of TU sema, and that would seem to be significantly more expensive, _and_ we might run into instantiation problems anyway in builds with -Werror-{tmo warning} invalidating instantiations in ways that impact semantics.

AI usage: I used claude to find many of the edge cases this PR fixes - the fixes are all written by me. I used claude to construct maybe half of the module and PCH tests, and then used those as a template for more. The biggest tell is whether Foo, Bar, Wibble, S, etc as identifiers.

### CODE REVIEW

File: clang/lib/AST/TypedMemoryInference.cpp [hunk 1]
```diff
@@ -148,6 +149,35 @@ std::string InferredAllocationType::describe(const ASTContext &Ctx) const {
...
 }
 
+const TypedMemoryAttr *CallExpr::getTypedMemoryAttribute() const {
+  const auto *Callee = dyn_cast_or_null<FunctionDecl>(getCalleeDecl());
+  if (!Callee)
+    return nullptr;
+  const TypedMemoryAttr *Attribute = Callee->getAttr<TypedMemoryAttr>();
+  // A call inside a template definition can bound to a call without a TMO
+  // attribute, but then the function can be redeclared with the attribute
+  // present, but prior to it being instantiated, so we'll look for that here.
```
Comment (ojhunt) [inline clang/lib/AST/TypedMemoryInference.cpp]: I just re-read this as the clunkiest comment ever, will address

File: clang/lib/Sema/SemaTypedMemory.cpp [hunk 2]
```diff
@@ -421,45 +583,34 @@ void Sema::handleTypedMemoryAttr(Decl *D, const ParsedAttr &AL) {
...
-        << InferredParam->getLocation();
+  std::optional<unsigned> ErrorKind;
+  if (!TargetDecl->hasWrittenPrototype())
+    ErrorKind = diag::TMOErrorKind::NoPrototype;
+  else if (isFunctionOrMethodVariadic(TargetDecl))
+    ErrorKind = diag::TMOErrorKind::Variadic;
+  else if (hasParameterPack(TargetDecl))
+    ErrorKind = diag::TMOErrorKind::ParameterPack;
+  else if (isInstanceMethod(TargetDecl))
+    ErrorKind = diag::TMOErrorKind::InstanceMethod;
```
Comment (t-rasmud) [inline clang/lib/Sema/SemaTypedMemory.cpp]: Do we also need to check `isOverloadedOperator` like in the SourceDecl case?

File: clang/lib/Sema/SemaTypedMemory.cpp [hunk 3]
```diff
@@ -338,56 +401,158 @@ bool Sema::checkTMOGetTypeDescriptor(QualType T, SourceLocation Loc,
...
+  // * K & R declarations have synthesized parameters that we can't reason
+  //   about.
+  // * Variadic functions and functions with packs: conservative but has not yet
+  //   proven to be a problem.
+  // * Instance methods: again conservative, as yet has not seemed to be a
+  //   problem.
+  // * Overloaded operators: intentionally not supported, because it results in
+  //   downstream sema weirdness due to different AST nodes being involved, and
+  //   has extremely questionable value.
+  if (!SourceDecl->hasWrittenPrototype() || isFunctionOrMethodVariadic(D) ||
```
Comment (t-rasmud) [inline clang/lib/Sema/SemaTypedMemory.cpp]: should this check for `hasPrototype()` if we want to account for inherited prototypes?

File: clang/lib/Sema/SemaTypedMemory.cpp (hunk 2 above)
Comment (ojhunt) [reply clang/lib/Sema/SemaTypedMemory.cpp]: - [x] I didn't _think_ so - but I've come up with something that might be able to trigger that path.

File: clang/lib/Sema/SemaTypedMemory.cpp (hunk 3 above)
Comment (ojhunt) [reply clang/lib/Sema/SemaTypedMemory.cpp]: - [x] That's something I wasn't sure about. Let me see if I can find a way to break this.

File: clang/lib/Sema/SemaTypedMemory.cpp (hunk 2 above)
Comment (ojhunt) [reply clang/lib/Sema/SemaTypedMemory.cpp]: I've refactored this code so that the validation is done by a shared lambda rather than allowing divergence

File: clang/lib/Sema/SemaTypedMemory.cpp (hunk 3 above)
Comment (ojhunt) [reply clang/lib/Sema/SemaTypedMemory.cpp]: No, because this was actually the case that led to me using hasWrittenPrototype :D

**Review by t-rasmud — APPROVED** [review]
LGTM!
